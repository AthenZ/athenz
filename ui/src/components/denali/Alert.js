/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';
import Icon from '../denali/icons/Icon';
import {
    colors,
    cssDropShadow,
    cssFontStyles,
    cssFontWeights,
} from '../denali/styles';
import { css, cx } from 'emotion';
import { rgba } from 'polished';

const TYPE_COLOR_MAP = {
    danger: 'statusDanger',
    info: 'statusInfo',
    success: 'statusSuccess',
    warning: 'statusWarning',
};

const TYPE_ICON_MAP = {
    danger: 'stop-warning',
    info: 'information-circle',
    success: 'check-circle',
    warning: 'warning',
};

// http://reactcommunity.org/react-modal/examples/set_app_element.html
ReactModal.setAppElement('body');

const modalOverlayClass = css`
    background-color: transparent;
    bottom: 0;
    left: 0;
    position: fixed;
    right: 0;
    top: 0;
    z-index: 5;
    label: modal-overlay;
`;

const makeAlertClass = (props) => css`
    ${cssDropShadow};
    align-items: flex-start;
    background: ${colors.white};
    border-radius: 4px;
    border-top: 4px solid ${colors[TYPE_COLOR_MAP[props.type]]};
    box-sizing: border-box;
    color: ${colors.black};
    display: flex;
    flex-flow: row nowrap;
    ${cssFontStyles.default};
    left: 50%;
    max-height: calc(100vh - 160px);
    min-width: 400px;
    outline: none;
    position: fixed;
    transform: translate(-50%, 0);
    width: 400px;
    label: alert;
    &.ReactModal__Content {
        opacity: 0;
        top: 0;
    }
    &.animated {
        transition: opacity 0.2s ease-out,
            top 0.3s cubic-bezier(0.175, 0.885, 0.32, 1.275);
    }
    &.ReactModal__Content--after-open {
        top: 80px;
        opacity: 1;
    }
    /* Animate fade out on close. This is currently not used as tests fail */
    &.ReactModal__Content--before-close {
        opacity: 0;
    }
    & .left-section {
        flex: 0 0 28px;
        padding: 14px;
        & .alert-icon {
            align-items: flex-start;
            display: flex;
            justify-content: center;
        }
    }
    & .right-section {
        display: flex;
        flex: 1 1 auto;
        flex-flow: column nowrap;
        & .title-row {
            align-items: center;
            display: flex;
            flex-flow: row nowrap;
            height: initial;
            justify-content: space-between;
            padding: 14px 14px 14px 0;
            & .title {
                ${cssFontWeights.bold};
            }
        }
        & .description {
            font-size: 14px;
            font-weight: 300;
            padding: 0 14px 14px 0;
        }
    }
`;

/**
 * Alert displays a dismissable modal near the top of the screen. It is meant to
 * show a short blurb of information. There are four
 * alert types:
 *
 * 1. Info
 * 2. Success
 * 3. Warning
 * 4. Danger
 *
 * Under the hood, Alert uses [`react-modal`](http://reactcommunity.org/react-modal/).
 * To prevent the page from scrolling when the modal is showing, add the
 * following to the app's global CSS:
 *
 * ```css
 * .ReactModal__Body--open {
 *   overflow: hidden;
 * }
 */
class Alert extends React.Component {
    // Taken from:
    // https://stackoverflow.com/questions/45678517/better-way-to-cleartimeout-in-componentwillunmount
    setTimer = (ms) => {
        // Remember the timer handle
        this.timerHandle = setTimeout(() => {
            this.timerHandle = null;
            this.props.onClose();
        }, ms);
    };

    clearTimer = () => {
        // Is our timer running?
        if (this.timerHandle) {
            clearTimeout(this.timerHandle);
            this.timerHandle = null;
        }
    };

    componentDidMount() {
        if (this.props.isOpen && this.props.duration) {
            this.setTimer(this.props.duration);
        }
    }

    componentDidUpdate(prevProps) {
        if (!prevProps.isOpen && this.props.isOpen && this.props.duration) {
            this.setTimer(this.props.duration);
        } else if (prevProps.isOpen && !this.props.isOpen) {
            this.clearTimer();
        }
    }

    componentWillUnmount() {
        this.clearTimer();
    }

    render() {
        const StatusIcon = TYPE_ICON_MAP[this.props.type];
        const alertClass = makeAlertClass(this.props);

        return (
            <ReactModal
                aria={{
                    labelledby: 'alert-title',
                    describedby: 'alert-description',
                }}
                className={cx(
                    alertClass,
                    'denali-alert',
                    `denali-alert-${this.props.type}`,
                    {
                        animated: !this.props.noanim,
                    },
                    this.props.className
                )}
                closeTimeoutMS={200}
                descriptionLabel={`${this.props.type} alert`}
                isOpen={this.props.isOpen}
                overlayClassName={modalOverlayClass}
                shouldCloseOnOverlayClick={false}
                onRequestClose={this.props.onClose}
            >
                <div className='left-section'>
                    <div className='alert-icon' data-testid='alert-icon'>
                        <Icon
                            icon={'check-circle'}
                            color={colors[TYPE_COLOR_MAP[this.props.type]]}
                            decorative={true}
                            size='28px'
                            verticalAlign='middle'
                        />
                    </div>
                </div>
                <div className='right-section'>
                    <div className='title-row'>
                        <div
                            className='title'
                            id='alert-title'
                            data-testid='alert-title'
                        >
                            {this.props.title}
                        </div>
                        <div
                            className='close-button'
                            onClick={this.props.onClose}
                        >
                            <Icon
                                color={rgba(colors.grey800, 0.4)}
                                colorHover={rgba(colors.grey800, 0.6)}
                                isLink
                                size='28px'
                                icon={'close'}
                            />
                        </div>
                    </div>
                    {this.props.description && (
                        <div
                            className='description'
                            id='alert-description'
                            data-testid='alert-description'
                        >
                            {this.props.description}
                        </div>
                    )}
                </div>
            </ReactModal>
        );
    }
}

Alert.propTypes = {
    /** Additonal class to apply to the outer div of the modal content */
    className: PropTypes.string,
    /** Optional description. Can be a string or JSX / component */
    description: PropTypes.any,
    /** Once open, call `onClose()` after N milliseconds */
    duration: PropTypes.number,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Toggle the alert on or off */
    isOpen: PropTypes.bool,
    /** Title of the alert */
    title: PropTypes.string.isRequired,
    /** Alert supports four status types */
    type: PropTypes.oneOf(['danger', 'info', 'success', 'warning']),
    /** Handler for closing the alert */
    onClose: PropTypes.func.isRequired,
};

Alert.defaultProps = {
    duration: 0,
    isOpen: false,
    noanim: false,
    type: 'info',
};

export default Alert;
