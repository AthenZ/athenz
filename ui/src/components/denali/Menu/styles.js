/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { css } from 'emotion';
import { cssFontStyles } from '../styles/fonts';
import { colors } from '../styles/colors';
import { cssDropShadow } from '../styles/drop-shadow';

export const makeCssPopperBox = (props) => css`
    ${cssDropShadow};
    ${cssFontStyles.default};
    background-color: ${colors.white};
    border: 2px solid ${colors.border};
    border-radius: 4px;
    color: ${colors.black};
    cursor: default;
    line-height: 1.4;
    margin: 0;
    max-width: ${props.basic ? '400px' : undefined};
    min-width: 20px;
    padding: ${props.basic ? '6px 10px' : 0};
    transition: ${!props.noanim && 'opacity 0.25s ease-in'};
    z-index: 999;
    label: denali-menu;
    /* Shift the popper to allow space for arrow */
    &[data-placement*='bottom'] {
        margin-top: 8px;
    }
    &[data-placement*='top'] {
        margin-bottom: 8px;
    }
    &[data-placement*='left'] {
        margin-right: 8px;
    }
    &[data-placement*='right'] {
        margin-left: 8px;
    }

    /**
   * Invisible padding between trigger and main content (around the arrow) so
   * user can slide from one to the other without the menu closing.
   */
    &::after {
        background: transparent;
        content: '';
        position: absolute;
    }
    &[data-placement*='top']::after {
        bottom: -12px;
        height: 12px;
        width: 100%;
    }
    &[data-placement*='bottom']::after {
        height: 12px;
        top: -12px;
        width: 100%;
    }
    &[data-placement*='left']::after {
        right: -12px;
        height: 100%;
        width: 12px;
    }
    &[data-placement*='right']::after {
        left: -12px;
        height: 100%;
        width: 12px;
    }
`;

export const cssArrow = css`
    height: 1rem;
    position: absolute;
    width: 1rem;
    label: menu-arrow;

    &::before,
    &::after {
        border-style: solid;
        content: '';
        display: block;
        height: 0;
        margin: auto;
        position: absolute;
        width: 0;
    }

    &[data-placement*='bottom'] {
        left: 0;
        margin-top: -6px;
        top: -8px;
        width: 1rem;
    }

    &[data-placement*='bottom']::before {
        border-color: transparent transparent ${colors.grey400} transparent;
        border-width: 0 8px 8px 8px;
        position: absolute;
        top: 4px;
    }

    &[data-placement*='bottom']::after {
        border-color: transparent transparent ${colors.white} transparent;
        border-width: 0 8px 8px 8px;
        top: 7px;
    }

    &[data-placement*='top'] {
        bottom: 0;
        height: 1rem;
        left: 0;
        margin-bottom: -1rem;
        width: 1rem;
    }

    &[data-placement*='top']::before {
        border-color: ${colors.grey400} transparent transparent transparent;
        border-width: 8px 8px 0 8px;
        position: absolute;
        top: 2px;
    }

    &[data-placement*='top']::after {
        border-color: ${colors.white} transparent transparent transparent;
        border-width: 8px 8px 0 8px;
        top: -1px;
    }

    &[data-placement*='right'] {
        height: 1rem;
        left: 0;
        margin-left: -1rem;
        width: 1rem;
    }

    &[data-placement*='right']::before {
        border-color: transparent ${colors.grey400} transparent transparent;
        border-width: 8px 8px 8px 0;
        left: 5px;
    }

    &[data-placement*='right']::after {
        border-color: transparent ${colors.white} transparent transparent;
        border-width: 8px 8px 8px 0;
        left: 8px;
        top: 0;
    }

    &[data-placement*='left'] {
        height: 1rem;
        margin-right: -1rem;
        right: 0;
        width: 1rem;
    }

    &[data-placement*='left']::before {
        border-color: transparent transparent transparent ${colors.grey400};
        border-width: 8px 0 8px 8px;
        left: 2px;
    }

    &[data-placement*='left']::after {
        border-color: transparent transparent transparent ${colors.white};
        border-width: 8px 0 8px 8px;
        left: -1px;
        top: 0;
    }
`;
