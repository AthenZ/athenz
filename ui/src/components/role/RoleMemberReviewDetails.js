/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import styled from '@emotion/styled';
import ReviewMembersRow from './ReviewMembersRow';
import Button from '../denali/Button';
import Color from '../denali/Color';
import InputLabel from '../denali/InputLabel';
import Input from '../denali/Input';
import RequestUtils from '../utils/RequestUtils';

const TDStyled = styled.td`
    background-color: ${(props) => props.color};
    text-align: ${(props) => props.align};
    padding: 30px 20px;
    vertical-align: middle;
    word-break: break-all;
`;
const TitleDiv = styled.div`
    font-size: 16px;
    font-weight: 600;
`;

const ReviewMembersContainerDiv = styled.div`
    margin: 0;
`;

const ReviewMembersSectionDiv = styled.div`
    margin: 20px;
`;

const ReviewMembersTable = styled.table`
    width: 100%;
    border-spacing: 0;
    display: table;
    border-collapse: separate;
    border-color: grey;
    table-layout: fixed;
`;

const TableHeadStyled = styled.th`
    text-align: ${(props) => props.align};
    border-bottom: 2px solid #d5d5d5;
    color: #9a9a9a;
    font-weight: 600;
    font-size: 0.8rem;
    padding-bottom: 5px;
    vertical-align: top;
    text-transform: uppercase;
    padding: 5px 0 5px 15px;
    word-break: break-all;
`;

const SubmitDiv = styled.div`
    margin-top: 10px;
`;

const SubmitTextSpan = styled.span`
    color: #9a9a9a;
`;

const StyledInputLabel = styled(InputLabel)`
    flex: 0 0 100px;
    margin-right: 2%;
`;

const StyledInput = styled(Input)`
    width: 500px;
`;

const ContentDiv = styled.div`
    flex: 1 1;
    margin-right: 10px;
`;

const StyledAnchor = styled.a`
    color: #3570f4;
    text-decoration: none;
    cursor: pointer;
`;

const StyledJustification = styled(Input)`
    width: 300px;
    margin-top: 5px;
`;

export default class RoleMemberReviewDetails extends React.Component {
    constructor(props) {
        super(props);
        this.api = this.props.api;
        this.submitReview = this.submitReview.bind(this);
        this.submitRoleMetaExpiry = this.submitRoleMetaExpiry.bind(this);
        this.updateRoleMeta = this.updateRoleMeta.bind(this);
        this.cancelRoleMetaUpdate = this.cancelRoleMetaUpdate.bind(this);
        this.onUpdate = this.onUpdate.bind(this);

        this.state = {
            list: this.props.members || [],
            memberExpiry: '',
            serviceExpiry: '',
            showUpdateRoleMeta: false,
            submittedReview: false,
        };
        this.loadRole();
    }

    loadRole() {
        this.props.api
            .getRole(this.props.domain, this.props.role, false, true, false)
            .then((role) => {
                if (role.trust != null) {
                    this.setState({
                        showTrustError: true,
                        errorMessage: `This is a delegated role. It needs to be reviewed in ${role.trust} domain.`,
                    });
                } else {
                    let members =
                        role.roleMembers &&
                        role.roleMembers.map((m) => m.memberName);
                    this.setState({
                        roleObj: role,
                        list: role.roleMembers || [],
                        showUpdateRoleMeta: !(
                            role.memberExpiryDays || role.serviceExpiryDays
                        ),
                        memberExpiry: role.memberExpiryDays,
                        serviceExpiry: role.serviceExpiryDays,
                        extendedMembers: new Set(members),
                        deletedMembers: new Set(),
                        submittedReview: false,
                    });
                }
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
            });
    }

    inputChanged(key, evt) {
        this.setState({ [key]: evt.target.value });
    }

    submitRoleMetaExpiry() {
        if (isNaN(this.state.memberExpiry) || isNaN(this.state.serviceExpiry)) {
            this.setState({
                errorMessage: 'Expiry days should be whole numbers. ',
            });
            return;
        }
        let roleMeta = {
            selfServe: this.state.roleObj.selfServe,
            reviewEnabled: this.state.roleObj.reviewEnabled,
        };
        if (!isNaN(this.state.memberExpiry)) {
            let memExpNum = parseInt(this.state.memberExpiry);
            if (memExpNum < 1) {
                this.setState({
                    errorMessage: 'Expiry days should be greater than 0. ',
                });
                return;
            } else {
                roleMeta.memberExpiryDays = memExpNum;
            }
        }
        if (!isNaN(this.state.serviceExpiry)) {
            let memExpNum = parseInt(this.state.serviceExpiry);
            if (memExpNum < 1) {
                this.setState({
                    errorMessage: 'Expiry days should be greater than 0. ',
                });
                return;
            } else {
                roleMeta.serviceExpiryDays = memExpNum;
            }
        }
        this.props.api
            .putRoleMeta(
                this.props.domain,
                this.props.role,
                roleMeta,
                'Added using Athenz UI',
                this.props._csrf
            )
            .then(() => {
                this.setState({
                    showUpdateRoleMeta: false,
                });
                this.loadRole();
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
            });
    }

    submitReview() {
        if (
            this.state.roleObj.roleMembers &&
            this.state.roleObj.roleMembers.length > 0
        ) {
            if (
                this.state.justification === undefined ||
                this.state.justification.trim() === ''
            ) {
                this.setState({
                    errorMessage:
                        'Justification is required to submit the review.',
                });
                return;
            }

            //construct role object from state
            let role = {
                name: this.props.role,
            };
            role.roleMembers = this.state.roleObj.roleMembers;
            role.roleMembers.forEach((m) => {
                if (this.state.deletedMembers.has(m.memberName)) {
                    m.active = false;
                }
                m.expiration = null;
                delete m.memberFullName; // memberFullName is not a valid property on the server
            });
            role.roleMembers = role.roleMembers.filter((m) => {
                if (
                    this.state.deletedMembers.has(m.memberName) ||
                    this.state.extendedMembers.has(m.memberName)
                ) {
                    return m;
                }
            });
            this.props.api
                .reviewRole(
                    this.props.domain,
                    this.props.role,
                    role,
                    this.state.justification,
                    this.props._csrf
                )
                .then(() => {
                    this.setState({
                        submittedReview: true,
                        errorMessage: null,
                    });
                    this.props.onUpdateSuccess(
                        `Successfully submitted the review for role ${this.props.role}`
                    );
                })
                .catch((err) => {
                    this.setState({
                        errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                    });
                });
        } else {
            this.props.onUpdateSuccess('There is nothing to review.');
        }
    }

    updateRoleMeta() {
        this.setState({ showUpdateRoleMeta: true });
    }

    cancelRoleMetaUpdate() {
        this.setState({ showUpdateRoleMeta: false });
    }

    onUpdate(key, value) {
        switch (key) {
            case 'delete':
                this.state.deletedMembers.add(value);
                this.state.extendedMembers.delete(value);
                break;
            case 'extend':
                this.state.extendedMembers.add(value);
                this.state.deletedMembers.delete(value);
                break;
            case 'no-action':
                this.state.extendedMembers.delete(value);
                this.state.deletedMembers.delete(value);
                break;
        }
    }

    getDefaultExpiryText() {
        let text = 'Current default settings are - ';
        if (this.state.roleObj && this.state.roleObj.memberExpiryDays) {
            text =
                text +
                'Member expiry: ' +
                this.state.roleObj.memberExpiryDays +
                ' days ';
        }
        if (this.state.roleObj && this.state.roleObj.serviceExpiryDays) {
            text =
                text +
                'Service expiry: ' +
                this.state.roleObj.serviceExpiryDays +
                ' days ';
        }

        const changeText = 'To change it, please click ';

        return (
            <SubmitTextSpan>
                {text}
                <br />
                {changeText}
                <StyledAnchor onClick={this.updateRoleMeta}>
                    {' '}
                    here{' '}
                </StyledAnchor>
            </SubmitTextSpan>
        );
    }

    getContent() {
        const left = 'left';
        let center = 'center';
        const rows =
            this.state.list && this.state.list.length > 0
                ? this.state.list
                      .sort((a, b) => {
                          return a.memberName.localeCompare(b.memberName);
                      })
                      .map((item, i) => {
                          let color = 'white';
                          return (
                              <ReviewMembersRow
                                  key={'role-review-' + i}
                                  idx={'role-review-' + i}
                                  details={item}
                                  role={this.props.role}
                                  color={color}
                                  onUpdate={this.onUpdate}
                                  submittedReview={this.state.submittedReview}
                              />
                          );
                      })
                : [];

        if (this.state.showTrustError) {
            return (
                <ReviewMembersContainerDiv>
                    <ContentDiv>
                        {this.state.errorMessage && (
                            <Color name={'red600'}>
                                {this.state.errorMessage}
                            </Color>
                        )}
                    </ContentDiv>
                </ReviewMembersContainerDiv>
            );
        }

        if (!this.state.showUpdateRoleMeta) {
            if (
                isNaN(this.state.memberExpiry) &&
                isNaN(this.state.serviceExpiry)
            ) {
                //user clicked on cancel without entering memberExpiry and serviceExpiry. We are going to
                // close the widget.
                {
                    this.props.onUpdateSuccess();
                }
            }

            return (
                <ReviewMembersContainerDiv>
                    <TitleDiv>REVIEW EXPIRING MEMBERS</TitleDiv>
                    <ReviewMembersSectionDiv data-testid='review-member-list'>
                        <ReviewMembersTable>
                            <thead>
                                <tr>
                                    <TableHeadStyled align={left}>
                                        MEMBER
                                    </TableHeadStyled>
                                    <TableHeadStyled align={left}>
                                        MEMBER NAME
                                    </TableHeadStyled>
                                    <TableHeadStyled align={left}>
                                        EXPIRATION DATE
                                    </TableHeadStyled>
                                    <TableHeadStyled align={center}>
                                        EXTEND
                                    </TableHeadStyled>
                                    <TableHeadStyled align={center}>
                                        NO ACTION
                                    </TableHeadStyled>
                                    <TableHeadStyled align={center}>
                                        DELETE
                                    </TableHeadStyled>
                                </tr>
                            </thead>
                            <tbody>
                                {rows}
                                <tr key='submit-review'>
                                    <td colSpan={2}>
                                        <StyledJustification
                                            id='justification'
                                            name='justification'
                                            value={
                                                this.state.justification
                                                    ? this.state.justification
                                                    : ''
                                            }
                                            onChange={this.inputChanged.bind(
                                                this,
                                                'justification'
                                            )}
                                            autoComplete={'off'}
                                            placeholder='Enter justification here'
                                        />
                                    </td>
                                    <td colSpan={1}>
                                        <SubmitDiv>
                                            <Button
                                                secondary={true}
                                                onClick={this.submitReview}
                                            >
                                                Submit Review
                                            </Button>
                                        </SubmitDiv>
                                    </td>
                                    <td colSpan={3}>
                                        {this.getDefaultExpiryText()}
                                    </td>
                                </tr>
                                <tr key='error-message'>
                                    <td colSpan={6}>
                                        {this.state.errorMessage && (
                                            <Color name={'red600'}>
                                                {this.state.errorMessage}
                                            </Color>
                                        )}
                                    </td>
                                </tr>
                            </tbody>
                        </ReviewMembersTable>
                    </ReviewMembersSectionDiv>
                </ReviewMembersContainerDiv>
            );
        } else {
            return (
                <ReviewMembersContainerDiv>
                    User principal and/or Service principal expiry days are
                    required before a role can be reviewed.
                    <div>
                        <StyledInputLabel htmlFor='role-meta-member-expiry'>
                            User Principal Expiry ( in Days )
                        </StyledInputLabel>
                        <ContentDiv>
                            <StyledInput
                                id='role-meta-member-expiry'
                                name='role-meta-member-expiry'
                                value={
                                    this.state.memberExpiry
                                        ? '' + this.state.memberExpiry
                                        : ''
                                }
                                onChange={(e) =>
                                    this.setState({
                                        memberExpiry: e.target.value,
                                    })
                                }
                            />
                        </ContentDiv>
                    </div>
                    <div>
                        <StyledInputLabel htmlFor='role-meta-service-expiry'>
                            Service Principal Expiry ( in Days )
                        </StyledInputLabel>
                        <ContentDiv>
                            <StyledInput
                                id='role-meta-service-expiry'
                                name='role-meta-service-expiry'
                                value={
                                    this.state.serviceExpiry
                                        ? '' + this.state.serviceExpiry
                                        : ''
                                }
                                onChange={(e) =>
                                    this.setState({
                                        serviceExpiry: e.target.value,
                                    })
                                }
                            />
                        </ContentDiv>
                    </div>
                    <SubmitDiv>
                        <Button
                            secondary={true}
                            onClick={this.submitRoleMetaExpiry}
                        >
                            Submit Expiry
                        </Button>
                        <Button
                            secondary={true}
                            onClick={this.cancelRoleMetaUpdate}
                        >
                            Cancel
                        </Button>
                    </SubmitDiv>
                    {this.state.errorMessage && (
                        <Color name={'red600'}>{this.state.errorMessage}</Color>
                    )}
                </ReviewMembersContainerDiv>
            );
        }
    }

    render() {
        let { color } = this.props;
        return (
            <TDStyled
                color={color}
                colSpan={8}
                data-testid='role-member-review-details'
            >
                {this.getContent()}
            </TDStyled>
        );
    }
}
