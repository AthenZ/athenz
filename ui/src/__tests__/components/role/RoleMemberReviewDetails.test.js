/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import RoleMemberReviewDetails from '../../../components/role/RoleMemberReviewDetails';
import { colors } from '../../../components/denali/styles';

describe('RoleMemberReviewDetails', () => {
    it('should render', () => {
        let color = colors.row;
        let domain = 'domain';
        let role = 'role';

        const api = {
            getRole(domain) {
                return new Promise((resolve, reject) => {
                    let role = {
                        roleMembers: [{ memberName: 'a' }, { memberName: 'b' }],
                        memberExpiryDays: 30,
                        serviceExpiryDays: 20,
                    };
                    resolve(['a, b']);
                });
            },
        };

        const { getByTestId } = render(
            <table>
                <tbody>
                    <tr>
                        <RoleMemberReviewDetails
                            color={color}
                            domain={domain}
                            role={role}
                            api={api}
                        />
                    </tr>
                </tbody>
            </table>
        );
        const auditLogList = getByTestId('review-member-list');

        expect(auditLogList).toMatchSnapshot();
    });
});
