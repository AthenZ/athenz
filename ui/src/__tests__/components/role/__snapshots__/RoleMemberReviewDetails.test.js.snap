// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoleMemberReviewDetails should render 1`] = `
.emotion-6 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  position: relative;
  width: 250px;
}

.emotion-6 input {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: rgba(53,112,244,0.05);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #303030;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: 36px;
  margin: 0;
  outline: none;
  padding: 0 1rem;
  text-align: left;
  width: 100%;
}

.emotion-6.animated input {
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out, border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out, border 0.2s ease-in-out;
}

.emotion-6 input::-webkit-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-6 input::-moz-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-6 input:-ms-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-6 input::placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-6 input.focused,
.emotion-6 input:active,
.emotion-6 input:focus {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #3570f4;
  color: #303030;
}

.emotion-6 input:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-6 .message {
  font-size: 12px;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

.emotion-6.error input,
.emotion-6.error input.focused,
.emotion-6.error input:active,
.emotion-6.error input:focus {
  border-bottom: 2px solid #d01111;
}

.emotion-6.error .message {
  color: #d01111;
}

.emotion-8 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 14px;
  line-height: 1;
  margin: 5px;
  outline: 0;
  padding: 10px 24px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: transparent;
  box-shadow: 0 0 0 1px #3697f2 inset;
  color: #3697f2;
}

.emotion-8:first-of-type {
  margin-left: 0;
}

.emotion-8:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-8:disabled {
  box-shadow: 0 0 0 1px rgba(48,48,48,0.1) inset;
}

.emotion-8:hover:not(:disabled) {
  background: #d7e2fd;
}

.emotion-13 {
  margin: 20px;
}

.emotion-12 {
  width: 100%;
  border-spacing: 0;
  display: table;
  border-collapse: separate;
  border-color: grey;
  table-layout: fixed;
}

.emotion-0 {
  text-align: left;
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.8rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 15px;
  word-break: break-all;
}

.emotion-3 {
  text-align: center;
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.8rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 15px;
  word-break: break-all;
}

.emotion-7 {
  width: 300px;
  margin-top: 5px;
}

.emotion-9 {
  margin-top: 10px;
}

.emotion-11 {
  color: #9a9a9a;
}

.emotion-10 {
  color: #3570f4;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

<div
  class="emotion-13"
  data-testid="review-member-list"
>
  <table
    class="emotion-12"
  >
    <thead>
      <tr>
        <th
          class="emotion-0"
        >
          MEMBER
        </th>
        <th
          class="emotion-0"
        >
          MEMBER NAME
        </th>
        <th
          class="emotion-0"
        >
          EXPIRATION DATE
        </th>
        <th
          class="emotion-3"
        >
          EXTEND
        </th>
        <th
          class="emotion-3"
        >
          NO ACTION
        </th>
        <th
          class="emotion-3"
        >
          DELETE
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colspan="2"
        >
          <div
            class="emotion-6 denali-input animated emotion-7"
            data-testid="input-wrapper"
          >
            <input
              autocomplete="off"
              data-testid="input-node"
              id="justification"
              name="justification"
              placeholder="Enter justification here"
              value=""
            />
            <div
              class="message"
              data-testid="message"
            />
          </div>
        </td>
        <td
          colspan="1"
        >
          <div
            class="emotion-9"
          >
            <button
              class="emotion-8 denali-button"
            >
              Submit Review
            </button>
          </div>
        </td>
        <td
          colspan="3"
        >
          <span
            class="emotion-11"
          >
            Current default settings are - 
            <br />
            To change it, please click 
            <a
              class="emotion-10"
            >
               
              here
               
            </a>
          </span>
        </td>
      </tr>
      <tr>
        <td
          colspan="6"
        />
      </tr>
    </tbody>
  </table>
</div>
`;
